AX: Optimize AXIsolatedObject::boolAttributeValue by matching AXProperty order with AXPropertyFlag order

twilco opened the pull request:
#### 280d982bb70fc0a69ed208fd5a45514ca8554dd3
<pre>
AX: Optimize AXIsolatedObject::boolAttributeValue by matching AXProperty order with AXPropertyFlag order
<a href="https://bugs.webkit.org/show_bug.cgi?id=294743">https://bugs.webkit.org/show_bug.cgi?id=294743</a>
<a href="https://rdar.apple.com/153866597">rdar://153866597</a>

Reviewed by Joshua Hoffman.

Prior to this commit, AXIsolatedObject::boolAttributeValue had 21 branches to map the given AXProperty to an AXPropertyFlag.
With this commit, we avoid all of these branches by making the property ordering in AXProperty and AXPropertyFlag match,
meaning we can cast between them, and quickly determine whether it is a property that has a corresponding flag (static_cast(property) &lt;= lastPropertyFlagIndex).

boolAttributeValue is one of our most called functions. Here is the before and after runtimes of boolAttributeValue
while holding VO-Right on the HTML spec page for ~2.5 mins.

Before: 398665000 calls, mean duration: 0.000023ms, total duration: 9177.080922ms
After: 398665000 calls, mean duration: 0.000012ms, total duration: 4964.236194ms

* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::boolAttributeValue const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:

Canonical link: <a href="https://commits.webkit.org/296447@main">https://commits.webkit.org/296447@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/637037d6eb0cb3c1bbdfca626a537eb27735190e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108530 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28191 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18615 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113739 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/58928 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/110493 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28880 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36743 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82425 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111478 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22915 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97757 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62861 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22332 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15897 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58454 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92285 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15948 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116860 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35584 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/26229 "Found 7 new test failures: http/tests/media/hls/hls-audio-tracks-language.html http/tests/media/hls/hls-webvtt-default.html http/tests/media/hls/hls-webvtt-style.html http/tests/media/hls/range-request-cross-origin.html http/tests/media/hls/track-in-band-multiple-cues.html http/tests/media/hls/track-webvtt-multitracks.html platform/mac/media/encrypted-media/fps-generateRequest.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91450 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/35957 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94029 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91251 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23257 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36144 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13912 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/31340 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35486 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/41021 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35197 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38547 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36876 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/637037d6eb0cb3c1bbdfca626a537eb27735190e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46979

twilco committed:
AX: Optimize AXIsolatedObject::boolAttributeValue by matching AXProperty order with AXPropertyFlag order
https://bugs.webkit.org/show_bug.cgi?id=294743
rdar://153866597

Reviewed by Joshua Hoffman.

Prior to this commit, AXIsolatedObject::boolAttributeValue had 21 branches to map the given AXProperty to an AXPropertyFlag.
With this commit, we avoid all of these branches by making the property ordering in AXProperty and AXPropertyFlag match,
meaning we can cast between them, and quickly determine whether it is a property that has a corresponding flag (static_cast(property) <= lastPropertyFlagIndex).

boolAttributeValue is one of our most called functions. Here is the before and after runtimes of boolAttributeValue
while holding VO-Right on the HTML spec page for ~2.5 mins.

Before: 398665000 calls, mean duration: 0.000023ms, total duration: 9177.080922ms
After: 398665000 calls, mean duration: 0.000012ms, total duration: 4964.236194ms

* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::boolAttributeValue const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:

Canonical link: https://commits.webkit.org/296447@main
